sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java

```diff
@@ -1036,27 +1040,82 @@ public PagedFlux<ConfigurationSetting> listConfigurationSettings(SettingSelector
-                    null,
-                    addTracingNamespace(context))
-                               .map(pagedResponse -> toConfigurationSettingWithPagedResponse(pagedResponse)))
+                () -> withContext(context -> {
+                    String firstPageETag = (matchConditionsList == null || matchConditionsList.isEmpty())
+                            ? null
+                            : matchConditionsList.get(0).getIfNoneMatch();
+                    return serviceClient.getKeyValuesSinglePageAsync(
+                            keyFilter,
+                            labelFilter,
+                            null,
+                            acceptDateTime,
+                            settingFields,
+                            null,
+                            null,
+                            firstPageETag,
+                            addTracingNamespace(context))
+                            .onErrorResume(
+                                HttpResponseException.class,
+                                (Function<Throwable, Mono<PagedResponse<KeyValue>>>) throwable -> {
+                                    HttpResponseException e = (HttpResponseException) throwable;
+                                    HttpResponse httpResponse = e.getResponse();
+
+                                    if (httpResponse.getStatusCode() == 304) {
+                                        String continuationToken = parseNextLink(httpResponse.getHeaderValue("link"));
+                                        return Mono.just(
+                                                new PagedResponseBase<>(
+                                                        httpResponse.getRequest(),
+                                                        httpResponse.getStatusCode(),
+                                                        httpResponse.getHeaders(),
+                                                        null,
+                                                        continuationToken,
+                                                        null));
+                                    }
+                                    return Mono.error(throwable);
+                                })
+                            .map(pagedResponse -> toConfigurationSettingWithPagedResponse(pagedResponse));
+                }),
+                nextLink -> withContext(context -> {
+                    int pageETagListSize = (matchConditionsList == null || matchConditionsList.isEmpty())
+                            ? 0
+                            : matchConditionsList.size();
+                    String nextPageETag = null;
+                    int pageETagIndexValue = pageETagIndex.get();
+                    if (pageETagIndexValue < pageETagListSize) {
+                        nextPageETag = matchConditionsList.get(pageETagIndexValue).getIfNoneMatch();
+                        pageETagIndex.set(pageETagIndexValue + 1);
+                    }
```

Comment:
Let's turn this into a method that takes the `AtomicInteger` and the list of `MatchConditions` and returns the ETag. This can be shared between the initial page retrieval and next page retrieval, will just need to start the AtomicInteger at 0.